autoware_reference_system/test/generate_reports.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2021 Apex.AI, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import sys
+
+from bokeh.io import export_png, output_file
+from bokeh.layouts import layout
+from bokeh.plotting import save, show
+
+import callback_duration
+import memory_usage
+
+from tracetools_analysis.loading import load_file
+from tracetools_analysis.processor import Processor
+from tracetools_analysis.processor.memory_usage import KernelMemoryUsageHandler
+from tracetools_analysis.processor.memory_usage import UserspaceMemoryUsageHandler
+from tracetools_analysis.processor.ros2 import Ros2Handler
+from tracetools_analysis.utils.memory_usage import MemoryUsageDataModelUtil
+from tracetools_analysis.utils.ros2 import Ros2DataModelUtil
+
+path = ''
+pwd = ''
+ros2_data_model = None
+memory_data_model = None
+callback_symbols = None
+ust_memory_usage_dfs = None
+kernel_memory_usage_dfs = None
+
+SIZE_SUMMARY = 800
+SIZE_SUBPLOT = 500
+
+
+def checkPath(p):
+    # make sure path ends in a `/`
+    if (p[-1] != '/'):
+        p += '/'
+    # make sure directory exists
+    if not os.path.isdir(p):
+        print('Given path does not exist: ' + p)
+        sys.exit()
+
+    global pwd
+    pwd = os.path.basename(os.path.normpath(path))
+    print(pwd)
+    return p
+
+
+def initTraceData():
+    events = load_file(path)
+    ros2_handler = Ros2Handler.process(events)
+    ust_memory_handler = UserspaceMemoryUsageHandler()
+    kernel_memory_handler = KernelMemoryUsageHandler()
+
+    global ros2_data_model
+    # callback duration
+    ros2_data_model = Ros2DataModelUtil(ros2_handler.data)
+    global callback_symbols
+    callback_symbols = ros2_data_model.get_callback_symbols()
+
+    # memory usage
+    ros2_handler = Ros2Handler()
+    mem_proc = Processor(
+        ros2_handler,
+        ust_memory_handler,
+        kernel_memory_handler
+    )
+    mem_proc.process(events)
```

Comment:
to cut down on processing time, you could use a single `Processor` for all the handlers. This way you'd read through the trace events only once (instead of once for `Ros2Handler.process(events)` and once for `mem_proc.process(events)`).

Reply:
@christophebedard  i did that originally but was seeing that `KeyError` so I tried it this way. ill switch it back though, good tip!